packages/SwingSet/src/kernel/kernel.js

```diff
@@ -286,21 +286,62 @@ export default function buildKernel(kernelEndowments, kernelOptions = {}) {
   function removeVatManager(vatID, reason) {
     const old = ephemeral.vats.get(vatID);
     ephemeral.vats.delete(vatID);
-    const err = reason ? Error(reason) : undefined;
-    old.notifyTermination(err); // XXX TODO: most of the places where notifyTermination gets passed don't need it
+    old.notifyTermination(reason); // XXX TODO: most of the places where notifyTermination gets passed don't need it
     return old.manager.shutdown();
   }
 
-  function terminateVat(vatID, reason) {
+  function removeVatManagerOnFailure(vatID, reason) {
+    const old = ephemeral.vats.get(vatID);
+    ephemeral.vats.delete(vatID);
+    old.notifyTerminationOnFailure(reason); // XXX TODO: most of the places where notifyTermination gets passed don't need it
+    return old.manager.shutdown();
+  }
+
+  function terminateVatOnFailure(vatID, reason) {
     if (kernelKeeper.getVatKeeper(vatID)) {
       const promisesToReject = kernelKeeper.cleanupAfterTerminatedVat(vatID);
       const err = reason ? makeError(reason) : VAT_TERMINATION_ERROR;
       for (const kpid of promisesToReject) {
         resolveToError(kpid, err, vatID);
       }
+      removeVatManagerOnFailure(vatID, reason).then(
+        () => console.error(`terminated vat as failure ${vatID}, "${reason}"`),
+        e =>
+          console.error(
+            `problem terminating vat as failure ${vatID}, "${reason}"`,
+            e,
+          ),
+      );
+    }
+  }
+
+  function resolveToSuccess(kpid, errorData, expectedDecider) {
+    insistCapData(errorData);
+    const p = kernelKeeper.getResolveablePromise(kpid, expectedDecider);
+    const { subscribers, queue } = p;
+    let idx = 0;
+    for (const dataSlot of errorData.slots) {
+      kernelKeeper.incrementRefCount(dataSlot, `reject|s${idx}`);
+      idx += 1;
+    }
+    kernelKeeper.fulfillKernelPromiseToData(kpid, errorData);
+    // we don't notify the decider; they already know
+    notifySubscribersAndQueue(kpid, expectedDecider, subscribers, queue);
+    if (pendingMessageResults.has(kpid)) {
+      notePendingMessageResolution(kpid, 'fulfilled', errorData);
+    }
+  }
+
+  function terminateVat(vatID, reason) {
+    if (kernelKeeper.getVatKeeper(vatID)) {
+      const promisesToReject = kernelKeeper.cleanupAfterTerminatedVat(vatID);
+      const cause = reason ? makeError(reason) : VAT_TERMINATION_ERROR;
+      for (const kpid of promisesToReject) {
+        resolveToSuccess(kpid, cause, vatID);
```

Comment:
Wait what? No, the orphaned promises must reject. Dead vats do not successfully resolve their pending promises just because the vat was deliberately killed with a non-Error `reason`.

We should talk about how much we expose about the reasons that the vat died to the callers. I think we should only ever reject these orphaned promises with a neutral VAT_TERMINATION_ERROR. The vat might have died because of metering, or because someone decided to kill it in a way that included some error message, but I don't think it's a good idea to copy that error out to every promise that the vat ever exposed. In fact I think that might create an unwanted communication channel between the holder of the admin facet and the holder of any result promise ever created by the vat.

I think this might be a case where we don't create parallel `terminateVat` and `terminateVatOnFailure` functions.. not sure. At least the `promisesToReject` should behave the same in both cases, if the factoring is such that `removeVatManager` has two versions, then we must figure out a way to call the right one.

But I think this is calling out for a different layout.

Reply:
I agree that if there are orphaned promises, we should reject them. I'm trying to get the promise created in vatAdminWrapper to resolve positively when the contract asks for a clean shutdown. (That's the promise that the contract's creator can get to find out when the vat finishes.) The path it takes is through here, so resolveToError() would cause that promise to be rejected.

I would have expected removeVatManager's `old.notifyTermination(reason)` to be the path that leads to that fulfillment, but when I move the for loop after `removeVatManager()` or drop the for loop entirely, the promise in vatAdminWrapper doesn't get resolved at all.
Is the queueToExport() in `dynamicVat.sendTerminationNotification()` causing it that delivery to be delayed too long?

Reply:
Brian pointed out that the promises can't chain through the ZCF vat. I refactored to make the donePromise available from `startInstance()` rather than from zcf.

Reply:
For reference: we found that the `adminFacet.done` promise was indeed being rejected, but the contract client was getting a different promise, also named `done`, from ZCF itself. ZCF was subscribed to `adminFacet.done` and had a `.then` which resolved/rejected its own "done" with the same value. However, since ZCF lives in the doomed vat, this `then` was never called (the vat is dead by the time `adminFacet.done`'s subscribers are notified). And since all of the dead vat's unresolved (orphaned) promises are rejected when the vat dies, the contract client saw the orphaned-promise rejection.

The fix was to give the client a promise which is decided by the Zoe vat (which survives), rather than the ZCF+contract vat (which dies). This could be exactly `adminFacet.done`, or a different promise which is chained to it, as long as that promise doesn't live in the ZCF+contract vat.

Note that if we had promise "forwarding", and ZCF did a `p.resolve(adminVat.done)`, then we could conceivably have the kernel update its tables to make the contract client now be subscribed to `adminVat.done` directly, instead of a promise that lives in the ZCF+contract vat. Then they could be notified directly+correctly even if the ZCF+contract vat dies, since that vat has now been cut out of the loop. I don't think we have enough information from `Promise` (and maybe not even `HandledPromise`) to detect forwarding, so I haven't worried too much about implementing it. Also, it opens up a question of whether forwarding is present or not, and I'm not sure how visible we should make that.